gdk-pixbuf/gdk-pixbuf-csource.1
.TH GDK-PIXBUF-CSOURCE 1 "04 Sep 2001"
.SH NAME
gdk-pixbuf-csource \- C code generation utility for GdkPixbuf images
.SH SYNOPSIS

\fBgdk-pixbuf-csource\fP [\fIoptions\fP] [\fIimage\fP]
.br
\fBgdk-pixbuf-csource\fP [\fIoptions\fP] \fI--build-list\fP [[\fIname image\fP]...]

.SH DESCRIPTION
\fBgdk-pixbuf-csource\fP is a small utility that generates C code containing
images, usefull for compiling images directly into programs.

.SH INVOCATION

\fBgdk-pixbuf-csource\fP either takes as input one image file name to generate code
for, or, using the \fI--build-list\fP option, a list of (\fIname\fP, \fIimage\fP)
pairs to generate code for a list of images into named variables.

.SS Options
.TP
\fI--stream
Generate pixbuf data stream (a single string containing a serialized
GdkPixdata structure in network byte order).

.TP
\fI--struct
Generate GdkPixdata structure (needs the GdkPixdata structure definition from
gdk-pixdata.h).

.TP
\fI--macros
Generate *_ROWSTRIDE, *_WIDTH, *_HEIGHT, *_BYTES_PER_PIXEL and
*_RLE_PIXEL_DATA or *_PIXEL_DATA macro definitions for the image.

.TP
\fI--rle
Enables run length encoding for the generated pixel data (default).

.TP
\fI--raw
Disables run length encoding for the generated pixel data (default).

.TP
\fI--extern
Generate extern symbols.

.TP
\fI--static
Generate static symbols (default).

.TP
\fI--decoder
Provide a *_RUN_LENGTH_DECODE(image_buf, rle_data, size, bpp) macro definition
to decode run-length encoded image data.

.TP
\fI--name=identifier
Specifies the identifier name (prefix) for the generated variable or
macros (usefull only if \fI--build-list\fP was not specified).

.TP
\fI--build-list
Enables (\fIname\fP, \fIimage\fP) pair parsing mode.

.TP
\fI-h, --help\fP 
Print brief help and exit.

.TP
\fI-v, --version\fP
Print version and exit.

.TP
\fI--g-fatal-warnings
Make warnings fatal (causes the program to abort).

.PP

.SH SEE ALSO
The \fIGdkPixbuf\fP documentation, shipped with the \fIGtk+\fP distribution,
available from \fIhttp://www.gtk.org\fP.

.SH BUGS 
None known yet.

.SH AUTHOR
.B gdk-pixbuf-csource
was written by Tim Janik <timj@gtk.org>.
.PP
This manual page was provided by Tim Janik <timj@gtk.org>.


NEWS
Overview of Changes in 4.24.1, 01-10-2026
=========================================

* Bugs fixed:
  - #8409 search-entry: Needs extra touch to show OSK although it has focus
    (Guido Günther)
  - #8419 text-input: Preedit gets cleared due to text-change-cause(1) (Guido
    Günther)
  - #8433 GtkSvgWidget can't parse '@media' if it is not on a new line (Matthias
    Clasen)
  - !10264 gtkmain: Split out debug code
  - !10367 gtkgstsink: release GL context on disposal (Pierre Berube)
  - !10369 Post-release version bump (Matthias Clasen)
  - !10372 Fix typo in CONTRIBUTING.md (Michael Fohqul)
  - !10385 texthandle: fix transformation issues & relative movement (Florian
    "sp1rit"​)
  - !10393 Fix typo ('witten' should be 'written') (ReinoutS)
  - !10395 searchbar: Show OSK when search-mode gets enabled (Guido Günther)
  - !10396 docs: Correct docs for allocation functions (Benjamin Otte)
  - !10398 textbuffer: Fix tag colors in accessible text attributes (Sean Rowe)
  - !10403 gtk/inscription: Refresh GtkInscription's PangoLayout when rooted
    (Christian Hergert)
  - !10409 text: Protect against invalid offset (Alessandro Astone)
  - !10414 macos: fix cursor keys (Arjan Molenaar)
  - !10421 various fixes (Benjamin Otte)
  - !10422 Norwegian Nynorsk translation. (Roy-Magne Mo)
  - !10423 gl: Create 2 fbos per frame instead of 1 per image (Benjamin Otte)
  - !10425 Update Norwegian nynorsk translation (Roy-Magne Mo)
  - !10429 textview: Make text handles visible after giving them a position
    (Carlos Garnacho)

* Translation updates:
  - Norwegian Nynorsk (Roy-Magne Mo)

Overview of Changes in 4.24.0, 11-09-2026
=========================================

* Bugs fixed:
  - #8195 Animated SVG frequently disappears (Matthias Clasen)
  - #8307 drop-performed class handlers read a phantom timestamp and send it in
    XdndDrop (Simeon Maxein)
  - #8313 macos: No gtk-macos-special equivalent for the Edit menu ("Start
    Dictation" / "Emoji & Symbols" missing with custom menubar)
  - #8354 GtkColumnView: sortable column headers expose no AT-SPI action, so
    sorting is unreachable for assistive technology
  - #8382 Media queries are not respected (Matthias Clasen)
  - #8390 svg: if the only animations in an svg are visibility animations
    elementa stay at initial value (Matthias Clasen)
  - #8400 Crash triggered by gtk_text_accessible_text_get_extents() (Matthias
    Clasen)
  - #8402 CUPS Secret Service authentication causes use-after-free in
    lookup_auth_info_cb() (Patrick Lambein)
  - !10132 Fix mismatches between parameters in four signal declarations and the
    corresponding class slots (Simeon Maxein)
  - !10245 gtkaccesskitcontext: Expose GActions as custom actions (Lukáš
    Tyrychtr)
  - !10265 macos: Default entries in Edit menu (Arjan Molenaar)
  - !10322 android: Remove empty_frame implementation (Benjamin Otte)
  - !10323 gtk: Repair action muxer subscriptions (Christian Hergert)
  - !10324 gsk: Allow conics with weight zero (Matthias Clasen)
  - !10328 CI: Switch glib CI back to a release (Benjamin Otte)
  - !10329 Update CI to be less likely to break (Benjamin Otte)
  - !10330 docs: Update release howto (Matthias Clasen)
  - !10331 Update Russian translation
  - !10333 Update Occitan translation
  - !10334 Update Norwegian Bokmål translation (Kjartan Maraas)
  - !10336 vulkan: Always update global buffer descriptor set (Benjamin Otte)
  - !10337 Update Serbian translation
  - !10338 Update French translation
  - !10339 svg: Fix run mode discrete (Matthias Clasen)
  - !10341 Update Basque translation (Asier Saratsua Garmendia)
  - !10343 ci: Download glib before configuring on MSYS2 (u9g)
  - !10344 Update Brazilian Portuguese translation
  - !10345 Update Spanish translation
  - !10346 memoryconvert: Pass the correct pointer address (Benjamin Otte)
  - !10349 Update Hungarian translation
  - !10350 Update Danish translation
  - !10353 Update Romanian translation
  - !10356 Update British English translation
  - !10359 curve: Add gsk_curve_get_extrema (Matthias Clasen)
  - !10360 Update Norwegian Nynorsk translation
  - !10361 Some frame clock fixes (Matthias Clasen)
  - !10365 gdkclipboard-x11: don't call a NULL converter (Marc-Antoine Perennou)

* Translation updates:
  - Basque (Asier Saratsua Garmendia)
  - Danish (Ask Hjorth Larsen)
  - English (United Kingdom) (Bruce Cowan)
  - French (Guillaume Bernard)
  - Hungarian (Balázs Úr)
  - Norwegian Bokmål (Kjartan Maraas)
  - Norwegian Nynorsk (Roy-Magne Mo)
  - Occitan (post 1500) (Quentin PAGÈS)
  - Portuguese (Brazil) (Rafael Coelho Costa)
  - Romanian (Antonio Marin)
  - Russian (Artur S0)
  - Serbian (Марко Костић)
  - Spanish (Daniel Mustieles García)

Overview of Changes in 4.23.4, 01-09-2026
=========================================

* Bugs fixed:
  - #7879 AT-SPI attempts to use /org/a11y/atspi/cache before it is ready w/gtk
    4 apps (Mike Gorse)
  - #8137 GtkListView misses items (Sergey Bugaev)
  - #8263 MacOS:  GMainContext uses stale poll array after re-entrant GPollFunc
  - #8293 macos: Drop to GTK4 app is not handled instantly
  - #8320 Issue: Fix Memory Leaks and Potential Crashes in GTK Components
    (Matthias Clasen)
  - #8330 GtkConstraintLayout: non-required inequality constraints release an
    error variable twice (Emmanuele Bassi)
  - #8334 Wayland: SIGSEGV in text_input_preedit_hint() when using IBus Chewing
    in GTK 4 applications (Takao Fujiwara)
  - #8343 macOS: Massive memory leak when unfocusing a window with an open popup
  - #8351 (GtkTextView) Cutting, pasting, or deleting text does not reset
    virtual cursor position (Greg Beauchesne)
  - #8355 GtkText's input interceptor logic is too eager (Carlos Garnacho)
  - #8357 gtk4-demo empty context menu on tabs (Matthias Clasen)
  - #8359 The menu bar focusing with F10 is wrong
  - #8360 GtkAccessibleTextInterface mismatch between interface documentation
    and its implementations (Matthias Clasen)

gtk-2-12/gdk-pixbuf/gdk-pixbuf-csource.c
/* Gdk-Pixbuf-CSource - GdkPixbuf based image CSource generator
 * Copyright (C) 1999, 2001 Tim Janik
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */
#include <config.h>

#include "../gtk/gtkversion.h"	/* versioning */
#include "gdk-pixbuf.h"
#include "gdk-pixdata.h"
#include <glib/gprintf.h>
#include <stdlib.h>
#include <string.h>


/* --- defines --- */
#undef	G_LOG_DOMAIN
#define	G_LOG_DOMAIN	"Gdk-Pixbuf-CSource"
#define PRG_NAME        "gdk-pixbuf-csource"
#define PKG_NAME        "Gtk+"
#define PKG_HTTP_HOME   "http://www.gtk.org"


/* --- prototypes --- */
static void	parse_args	(gint    *argc_p,
				 gchar ***argv_p);
static void	print_blurb	(FILE    *bout,
				 gboolean print_help);


/* --- variables --- */
static guint    gen_type = GDK_PIXDATA_DUMP_PIXDATA_STREAM;
static guint    gen_ctype = GDK_PIXDATA_DUMP_GTYPES | GDK_PIXDATA_DUMP_STATIC | GDK_PIXDATA_DUMP_CONST;
static gboolean use_rle = TRUE;
static gboolean with_decoder = FALSE;
static gchar   *image_name = "my_pixbuf";
static gboolean	build_list = FALSE;


/* --- functions --- */
static void
print_csource (FILE *f_out,
	       GdkPixbuf *pixbuf)
{
  GdkPixdata pixdata;
  gpointer free_me;
  GString *gstring;

  free_me = gdk_pixdata_from_pixbuf (&pixdata, pixbuf, use_rle);
  gstring = gdk_pixdata_to_csource (&pixdata, image_name,
				    gen_type | gen_ctype |
				    (with_decoder ? GDK_PIXDATA_DUMP_RLE_DECODER : 0));

  g_fprintf (f_out, "%s\n", gstring->str);

  g_free (free_me);
}

int
main (int   argc,
      char *argv[])
{
  GdkPixbuf *pixbuf;
  GError *error = NULL;
  gchar *infilename;

  /* initialize glib/GdkPixbuf */
  g_type_init ();

  /* parse args and do fast exits */
  parse_args (&argc, &argv);

  if (!build_list)
    {
      if (argc != 2)
	{
	  print_blurb (stderr, TRUE);
	  return 1;
	}
      
#ifdef G_OS_WIN32
      infilename = g_locale_to_utf8 (argv[1], -1, NULL, NULL, NULL);
#else
      infilename = argv[1];
#endif

      pixbuf = gdk_pixbuf_new_from_file (infilename, &error);
      if (!pixbuf)
	{
	  g_fprintf (stderr, "failed to load \"%s\": %s\n",
		   argv[1],
		   error->message);
	  g_error_free (error);
	  return 1;
	}
      
      print_csource (stdout, pixbuf);
      g_object_unref (pixbuf);
    }
  else /* parse name, file pairs */
    {
      gchar **p = argv + 1;
      guint j = argc - 1;
      gboolean toggle = FALSE;

      while (j--)
	{
#ifdef G_OS_WIN32
	  infilename = g_locale_to_utf8 (*p, -1, NULL, NULL, NULL);
#else
	  infilename = *p;
#endif

	  if (!toggle)
	    {
	      image_name = infilename;
	      p++;
	    }
	  else
	    {
	      pixbuf = gdk_pixbuf_new_from_file (infilename, &error);
	      if (!pixbuf)
		{
		  g_fprintf (stderr, "failed to load \"%s\": %s\n",
			   *p,
			   error->message);
		  g_error_free (error);
		  return 1;
		}
	      print_csource (stdout, pixbuf);
	      g_object_unref (pixbuf);
	      p++;
	    }
	  toggle = !toggle;
	}
    }
  
  return 0;
}

static void
parse_args (gint    *argc_p,
	    gchar ***argv_p)
{
  guint argc = *argc_p;
  gchar **argv = *argv_p;
  guint i, e;

  for (i = 1; i < argc; i++)
    {
      if (strcmp ("--macros", argv[i]) == 0)
	{
	  gen_type = GDK_PIXDATA_DUMP_MACROS;
	  argv[i] = NULL;
	}
      else if (strcmp ("--struct", argv[i]) == 0)
	{
	  gen_type = GDK_PIXDATA_DUMP_PIXDATA_STRUCT;
	  argv[i] = NULL;
	}
      else if (strcmp ("--stream", argv[i]) == 0)
	{
	  gen_type = GDK_PIXDATA_DUMP_PIXDATA_STREAM;
	  argv[i] = NULL;
	}
      else if (strcmp ("--rle", argv[i]) == 0)
	{
	  use_rle = TRUE;
	  argv[i] = NULL;
	}
      else if (strcmp ("--raw", argv[i]) == 0)
	{
	  use_rle = FALSE;
	  argv[i] = NULL;
	}
      else if (strcmp ("--extern", argv[i]) == 0)
	{
	  gen_ctype &= ~GDK_PIXDATA_DUMP_STATIC;
	  argv[i] = NULL;
	}
      else if (strcmp ("--static", argv[i]) == 0)
	{
	  gen_ctype |= GDK_PIXDATA_DUMP_STATIC;
	  argv[i] = NULL;
	}
      else if (strcmp ("--decoder", argv[i]) == 0)
	{
	  with_decoder = TRUE;
	  argv[i] = NULL;
	}
      else if ((strcmp ("--name", argv[i]) == 0) ||
	       (strncmp ("--name=", argv[i], 7) == 0))
	{
	  gchar *equal = argv[i] + 6;

	  if (*equal == '=')
	    image_name = g_strdup (equal + 1);
	  else if (i + 1 < argc)
	    {
	      image_name = g_strdup (argv[i + 1]);
	      argv[i] = NULL;
	      i += 1;
	    }
	  argv[i] = NULL;
	}
      else if (strcmp ("--build-list", argv[i]) == 0)
	{
	  build_list = TRUE;
	  argv[i] = NULL;
	}
      else if (strcmp ("-h", argv[i]) == 0 ||
	       strcmp ("--help", argv[i]) == 0)
	{
	  print_blurb (stderr, TRUE);
	  argv[i] = NULL;
	  exit (0);
	}
      else if (strcmp ("-v", argv[i]) == 0 ||
	       strcmp ("--version", argv[i]) == 0)
	{
	  print_blurb (stderr, FALSE);
	  argv[i] = NULL;
	  exit (0);
	}
      else if (strcmp (argv[i], "--g-fatal-warnings") == 0)
	{
	  GLogLevelFlags fatal_mask;

	  fatal_mask = g_log_set_always_fatal (G_LOG_FATAL_MASK);
	  fatal_mask |= G_LOG_LEVEL_WARNING | G_LOG_LEVEL_CRITICAL;
	  g_log_set_always_fatal (fatal_mask);

	  argv[i] = NULL;
	}
    }

  e = 0;
  for (i = 1; i < argc; i++)
    {
      if (e)
	{
	  if (argv[i])
	    {
	      argv[e++] = argv[i];
	      argv[i] = NULL;
	    }
	}
      else if (!argv[i])
	e = i;
    }
  if (e)
    *argc_p = e;
}

static void
print_blurb (FILE    *bout,
	     gboolean print_help)
{
  if (!print_help)
    {
      g_fprintf (bout, "%s version ", PRG_NAME);
      g_fprintf (bout, "%d.%d.%d", GTK_MAJOR_VERSION, GTK_MINOR_VERSION, GTK_MICRO_VERSION);
      g_fprintf (bout, "\n");
      g_fprintf (bout, "%s comes with ABSOLUTELY NO WARRANTY.\n", PRG_NAME);
      g_fprintf (bout, "You may redistribute copies of %s under the terms of\n", PRG_NAME);
      g_fprintf (bout, "the GNU Lesser General Public License which can be found in the\n");
      g_fprintf (bout, "%s source package. Sources, examples and contact\n", PKG_NAME);
      g_fprintf (bout, "information are available at %s\n", PKG_HTTP_HOME);
    }
  else
    {
      g_fprintf (bout, "Usage: %s [options] [image]\n", PRG_NAME);
      g_fprintf (bout, "       %s [options] --build-list [[name image]...]\n", PRG_NAME);
      g_fprintf (bout, "  --stream                   generate pixbuf data stream\n");
      g_fprintf (bout, "  --struct                   generate GdkPixdata structure\n");
      g_fprintf (bout, "  --macros                   generate image size/pixel macros\n");
      g_fprintf (bout, "  --rle                      use one byte run-length-encoding\n");
      g_fprintf (bout, "  --raw                      provide raw image data copy\n");
      g_fprintf (bout, "  --extern                   generate extern symbols\n");
      g_fprintf (bout, "  --static                   generate static symbols\n");
      g_fprintf (bout, "  --decoder                  provide rle decoder\n");
      g_fprintf (bout, "  --name=identifier          C macro/variable name\n");
      g_fprintf (bout, "  --build-list               parse (name, image) pairs\n");
      g_fprintf (bout, "  -h, --help                 show this help message\n");
      g_fprintf (bout, "  -v, --version              print version informations\n");
      g_fprintf (bout, "  --g-fatal-warnings         make warnings fatal (abort)\n");
    }
}

